From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Daniel Jurgens <danielj@nvidia.com>
Date: Sat, 10 Jan 2026 04:10:49 +0000
Subject: [PATCH 120/254] NVIDIA: SAUCE: tegra_hv_ivc: Defer setting the IRQ
 until reserve

Not all queues are IVC anymore. Defer getting the IRQ data until reserve
time.

Bug 5628239

Change-Id: I3088ebd447b84b669e7aa896ef260efe66551623
Signed-off-by: Daniel Jurgens <danielj@nvidia.com>
Reviewed-on: https://git-master.nvidia.com/r/c/3rdparty/canonical/linux-noble/+/3532109
Reviewed-by: svcacv <svcacv@nvidia.com>
GVS: buildbot_gerritrpt <buildbot_gerritrpt@nvidia.com>
Reviewed-by: Manish Bhardwaj <mbhardwaj@nvidia.com>
Reviewed-by: Brad Griffis <bgriffis@nvidia.com>
Tested-by: Manish Bhardwaj <mbhardwaj@nvidia.com>
(cherry picked from commit fb8a9a8d588206807eab8af3560aafbbd4ff7f75)

diff --git a/drivers/virt/tegra/tegra_hv_ivc.c b/drivers/virt/tegra/tegra_hv_ivc.c
index b08e2620a5a30..b6155fc48a6f5 100644
--- a/drivers/virt/tegra/tegra_hv_ivc.c
+++ b/drivers/virt/tegra/tegra_hv_ivc.c
@@ -103,7 +103,6 @@ static int tegra_hv_add_ivc(struct tegra_hv_data *hvd,
 	int rx_first;
 	struct iosys_map rx_map, tx_map;
 	uint32_t i;
-	struct irq_data *d;
 	uint64_t va_offset;
 
 	ivc = &hvd->ivc_devs[qd->id];
@@ -163,17 +162,7 @@ static int tegra_hv_add_ivc(struct tegra_hv_data *hvd,
 	if (ret < 0)
 		return -EINVAL;
 
-	ivc->irq = of_irq_get(hvd->dev, index);
-	if (ivc->irq < 0) {
-		ERR("Unable to get irq for ivc%u\n", qd->id);
-		return ivc->irq;
-	}
-	d = irq_get_irq_data(ivc->irq);
-	if (!d) {
-		ERR("Failed to get data for irq %d (ivc%u)\n", ivc->irq,
-				qd->id);
-		return -ENODEV;
-	}
+	ivc->queue_index = index;
 
 	if (qd->msi_ipa != 0U) {
 		if (WARN_ON(ivc_notify.msi_region_size == 0UL))
@@ -690,6 +679,19 @@ struct tegra_hv_ivc_cookie *tegra_hv_ivc_reserve(struct device_node *dn,
 	if (ret != 0)
 		return ERR_PTR(ret);
 
+	/* Get IRQ from device tree on first reserve */
+	if (ivc->irq == 0) {
+		ret = of_irq_get(hvd->dev, ivc->queue_index);
+		if (ret <= 0) {
+			ERR("Unable to get irq for ivc%u\n", ivc->qd->id);
+			mutex_lock(&ivc->lock);
+			ivc->reserved = 0;
+			mutex_unlock(&ivc->lock);
+			return ERR_PTR(ret);
+		}
+		ivc->irq = ret;
+	}
+
 	ivc->cookie_ops = ops;
 
 	ivck = &ivc->cookie;
-- 
2.43.0

